stdlib/Profile/src/memory.jl

```diff
@@ -0,0 +1,348 @@
+        StackFrame[],
+        UInt64(a.allocsz),
+        a.tag,
+
+        # We initialize this as `nothing` because we don't know if this
+        # object is actually still live or not.
+        nothing,
+    )
+end
+
+"""
+    close_AI(a::AllocationInfo, d::allocation_info_t, dealloc_stacktrace)
+
+Mark a previously opened `AllocationInfo` as closed, denoting that we know both the
+allocation and deallocation points for this chunk of memory.
+"""
+function close_AI(a::AllocationInfo, d::allocation_info_t, dealloc_stacktrace)
+    # Ensure that we're dealing with the same memory location
+    @assert a.address == UInt(d.address) "Attempting to close different memory locations! ($(a.address) != $(UInt(d.address)))"
+    @assert (a.tag | closing_tag) == d.tag "Attempting to close memory from different allocators/domains! ($(a.tag | closing_tag) != $(d.tag))"
+
+    # If a_type is still C_NULL, just set the type as `nothing`:
+    if d.T == C_NULL
+        d_type = nothing
+    else
+        d_type = unsafe_pointer_to_objref(d.T)
+    end
+
+    nstr(x) = x === nothing ? "nothing" : string(x)
+    @assert ((d_type == nothing || a.T == nothing) || d_type == a.T) "Mismatched types! ($(nstr(d_type)) != $(nstr(a.T)))"
+
+    return AllocationInfo(
+        a.address,
+        something(a.T, d_type, Some(nothing)),
+        a.alloc_time,
+        d.time,
+        a.alloc_stacktrace,
+        dealloc_stacktrace,
+        a.allocsz,
+        a.tag,
+        nothing,
+    )
+end
+
+macro memprofile(ex)
+    return quote
+        try
+            # Run a GC before we start profiling, to eliminate "ghost" chunks
+            Base.GC.gc()
+            start_memprofile()
+
+            # Explicit `begin/end` to help GC scope resolution
+            begin
+                $(esc(ex))
+            end
+        finally
+            # Run a GC afterwards to eliminate things that aren't actually leaked,
+            # they just weren't cleaned up before we stopped profiling.
+            Base.GC.gc()
+            stop_memprofile()
+        end
+    end
+end
+
+start_memprofile() = ccall(:jl_memprofile_start, Cvoid, ())
+stop_memprofile() = ccall(:jl_memprofile_stop, Cvoid, ())
+
+get_memprofile_bt_data() = ccall(:jl_memprofile_get_bt_data, Ptr{UInt}, ())
+get_memprofile_bt_data_len() = convert(Int, ccall(:jl_memprofile_len_bt_data, Csize_t, ()))
+get_memprofile_bt_data_maxlen() = convert(Int, ccall(:jl_memprofile_maxlen_bt_data, Csize_t, ()))
+
+get_memprofile_alloc_data() = ccall(:jl_memprofile_get_alloc_data, Ptr{allocation_info_t}, ())
+get_memprofile_alloc_data_len() = convert(Int, ccall(:jl_memprofile_len_alloc_data, Csize_t, ()))
+get_memprofile_alloc_data_maxlen() = convert(Int, ccall(:jl_memprofile_maxlen_alloc_data, Csize_t, ()))
+
+get_memprofile_overflow() = ccall(:jl_memprofile_overflow, Cint, ())
+get_memprofile_tag_filter() = UInt16(ccall(:jl_memprofile_tag_filter, Cint, ()))
+
+"""
+    read_and_coalesce_memprofile_data()
```

Comment:
As an API this feels awkward. The name suggests too much is being crammed into one function. If some users will only want detail on the "open" allocations, then the time spent constructing stacktraces for all the closed events will be a huge unnecessary load. Might be best to (1) parcel into types, (2) look for duplicates in the backtraces, and (3) look up only unique backtraces.